src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java

```diff
@@ -931,14 +939,15 @@ int doFinal(byte[] input, int inputOffset, int inputLen, byte[] output,
             // content of cipher has to be preserved in order for
             // users to retry the call with a larger buffer in the
             // case of ShortBufferException.
-            if (outputCapacity < estOutSize) {
+            if (output.length - outputOffset < estOutSize) {
                 cipher.save();
+                // create temporary output buffer if the estimated size is larger
+                // than the user-provided buffer.
+                internalOutput = new byte[estOutSize];
+                offset = 0;
             }
-            // create temporary output buffer so that only "real"
-            // data bytes are passed to user's output buffer.
-            outWithPadding = new byte[estOutSize];
```

Comment:
This change is somewhat dangerous. When using the supplied output buffer directly, you may corrupt its content w/ padding bytes. This optimization should only be applied when no padding is involved. In addition, input and output can point to the same buffer with all sorts of index combinations, i.e. inOfs == outOfs, inOfs < outOfs, inOfs > outOfs. With "outWithPadding" approach, no need to check. However, for "internalOutput", data corruption may happen. This kind of problem can be very hard to diagnose. Have to be very very careful here as this code may impact all ciphers...

Reply:
- I do not understand where the corruption comes from.  The user provides a buffer that output is suppose to be placed into, what could it be corrupting?  The existing tests (SameBuffer, in particular) works fine with this and the ByteBuffer calls.  I spent a lot of time trying to get those same buffer tests to pass.  
- It was my expectation that checkOutputCapacity() is making sure all the inOfs ==<> outOfs are going to work. Does that not catch all cases?
- outWithPadding" is excessive because it doubles the allocation for every operation followed by a copy to the original buffer, even if the original buffer was adequate.  I'm ok with doing the extra alloc & copy in certain situations, but not everytime.  Can you be more specific what things may fail that we don't already check for in the regression tests?

Reply:
I wrote a whole new tests to exercise all the buffers with and without offsets.  Hopefully that covers the concern.  The test found 4 bugs..

Reply:
Well, existing tests may not catch/check all error cases. Especially, in the past, all calls w/ ByteBuffer inputs are converted to calls w/ byte[] inputs. Thus, testing is focused on byte[] scenarios. In addition, since for decryption, internal buffering is done, there may be no test checking if padding bytes are written to the output buffer. Please see my comment follows.

Reply:
>     * It was my expectation that checkOutputCapacity() is making sure all the inOfs ==<> outOfs are going to work. Does that not catch all cases?
checkOutputCapacity() as the name has, only changes that the output buffer is large enough.

>     * outWithPadding" is excessive because it doubles the allocation for every operation followed by a copy to the original buffer, even if the original buffer was adequate.  I'm ok with doing the extra alloc & copy in certain situations, but not everytime.  Can you be more specific what things may fail that we don't already check for in the regression tests?

For example,
1) various input/output offset combinations, e.g. inOfs <,=,> outOfs, 
2) Given a output buffer of 200-byte and outOfs == 0, assuming the returned decrypted data length is 160 bytes, the bytes from index 160 and onward should not be overwritten. GCM has no padding, so you may not notice any difference if this is what you are testing. This is generic CipherCore code and changes impact all ciphers.

Reply:
checkOutputCapacity:  Yes.. The method includes the offsets for the output buffer, which I believe would verify that the output area in the buffer with offsets is large enough.

outWithPadding:  I understand the situation and I am assuming there are tests that cover this case.  Given it's a generic situation.

Reply:
Have you tested the outWithPadding situation? Given that the existing impl only write out the final result, I don't think you can assume that existing tests cover it. I have wrote a simple test to check it if you have not done so, can you try it out to be sure?

```
import java.io.PrintStream;
import java.util.*;
import java.security.*;
import java.security.spec.*;

import javax.crypto.*;
import javax.crypto.spec.*;

public class TestDoFinal {

    private static String ALGO = "AES";
    private static int BLK_SIZE = 16;

    public static void main(String args[]) throws Exception {

        byte[] in = new byte[32];
        Arrays.fill(in, (byte)8);
        KeyGenerator kg = KeyGenerator.getInstance(ALGO, "SunJCE");
        SecretKey skey = kg.generateKey();
        Cipher ci = Cipher.getInstance(ALGO + "/CBC/PKCS5Padding", "SunJCE");
        ci.init(Cipher.ENCRYPT_MODE, skey);
        int inLen = in.length - BLK_SIZE;
        byte[] out = ci.doFinal(in, 0, inLen);
        System.out.println("=> enc " + inLen + " bytes, ret " +
                (out == null? "null":(out.length + " byte")));

        AlgorithmParameters param = ci.getParameters();
        ci.init(Cipher.DECRYPT_MODE, skey, param);
        int rLen = ci.doFinal(out, 0, out.length, in);
        System.out.println("=> dec " + out.length + " bytes, ret " +
                rLen + " byte");
        // check if more than rLen bytes are written into 'in'
        for (int j = rLen; j < in.length; j++) {
            if (in[j] != (byte)8) {
                throw new Exception("Value check failed at index " + j);
            }
        }
        System.out.println("Test Passed");
    }
}
```

Reply:
I tried to fix this, and I did for this test, but there other situations with update() that weren't working.  It would take some reworking of a few common methods during the doFinal process to handle this right.  I'm going to put an 'if()" so non-GCM modes create a new buffer like it did before.  It was a "nice to have' for this rfe that can be done with future work for other mode optimizations.